### ISSUE TITLE
Keywords reserved cannot be detected with PHP 5.3 platform

### ISSUE THREAD

**Issue by llaville** [monque/PHP-Migration#4]
When your run on a PHP 5.3 plaform to detect if keywords reserved are used. E.g

```
php phpmig -s to54 /path/to/keywords_reserved.php
```

With source file (keywords_reserved.php)

``` php
<?php

class Trait
{
    const FOO = "foo";
}

class finally
{
    function __construct()
    {
        $trait = new Trait;

        $foo = Trait::FOO;
    }
}
```

You got following error

```
File: keywords_reserved.php
--------------------------------------------------------------------------------
    3 | PARSE      | * | NONE | Syntax error, unexpected T_TRAIT, expecting T_STRING on line 3
--------------------------------------------------------------------------------
```

Reason is that PHP-Parser is not yet able to catch such condition, until now. I've proposed an improvement https://github.com/nikic/PHP-Parser/issues/202

And soon, I'll propose a migration analyser on my php-compatinfo project, based on Sniff pattern. 
See https://github.com/llaville/php-compat-info/issues/202
I must gave you a credit, because this future analyser (that is available to detect even keywords reserved) is due to a deeper analysis of your project architecture.

**Comment by llaville** [comment]
You may be perharps interested by the presentation I wrote https://github.com/llaville/php-compat-info/issues/202#issuecomment-106249945 to explain how I solved this issue on my project.

**Comment by monque** [comment]
Your comments evokes my memory of developing this project...

I remember I have 3 solutions about the same problem **how to handle reserved keywords ?**
- Using lexer in PHP 5.3, `trait` is handled as a `T_STRING`
  - pros: reserved keywords will be checked as expected.
  - cons: parse failed if real `trait` appeared.
- Using emulative lexel, `trait` is handled as a `T_TRAIT`
  - pros: all new features (`trait`, `callable` ...) will be checked.
  - cons: parse failed if `trait` as a `identifier` (class name, function name...). (**typical error I want to check**)
- Additional process for reserved keywords: parse 2 times, first for reserved keywords, second for syntax
  - cons: complex...

After discussion, conclusion is that users want to check **incompatible** things and make sure his code runing on new version of PHP.

So the second solution meets requirement, It'll emit a _Syntax error_ if a reserved keyword appeared in wrong place, and all new features will be checked.

And the most important is, that's just simple, and do all we needs.

**Comment by llaville** [comment]
Don't forget to have a look on my new version of Emulative Lexer ( https://github.com/llaville/php-reflect/commit/c8281072da0980c5e7b7bdfb72cc354d93e8144c) that solved the problem

**Comment by llaville** [comment]
@monque Did you have time to have a look on my new Emulative lexer copy (see previous comment).

Since then, I've begun to write lot of sniffs and especially the [KeyWordReservedSniff](https://github.com/llaville/php-compat-info/blob/migration_analyser/src/Bartlett/CompatInfo/Sniffs/PHP/KeywordReservedSniff.php) that checks for keywords uses. It works fine for me on all PHP versions

**Comment by llaville** [comment]
@monque Be sure to pick the latest php-reflect version on sniffer branch. See last commit https://github.com/llaville/php-reflect/commit/4be20c38f8b11ada5b1ef7ddf549581c7ecd060a that fixed (I hope) the last detection issue

**Comment by monque** [comment]
@llaville This file [KeyWordReservedSniff](https://github.com/llaville/php-compat-info/blob/migration_analyser/src/Bartlett/CompatInfo/Sniffs/PHP/KeywordReservedSniff.php) makes a perfect solution.

Thanks for your contribution, I'm too busy to develop new feature..., so sorry for this.

Will you add all these "sniffs" for version compatibility to `php-compat-info` ?

**Comment by llaville** [comment]
@monque For next version 4.4.0 of php-compatinfo the `compatibility` analyser will be kept as actual. I planned to rewrite it usign sniffs.
But I'll introduced in the same time, a new `migration` analyser that will used all sniffs as currently defined in php-compatinfo `migration_analyser` branch.

**Comment by llaville** [comment]
I'll closed this issue, because php-compatinfo 4.5.x is the last version that will support PHP 5.3 and my version of Emulative Lexer to detect keywords.
Next major version that will come soon now ( 5.0 ) will support PHP 5.4 or above, and then dropped this version of Emulative Lexer.
